pkgs/development/python-modules/nanobind/default.nix

```diff
@@ -0,0 +1,54 @@
+{ lib, buildPythonPackage, pythonOlder, fetchFromGitHub, cmake, eigen, ninja
+, scikit-build, pytestCheckHook, numpy, scipy, torch, jax, jaxlib, tensorflow
+, setuptools }:
+buildPythonPackage rec {
+  pname = "nanobind";
+  version = "1.8.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "wjakob";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-MkbdjUL9ZC/JVBCSCgNLqTObISczPhJtDKKOSa85gmA=";
+    fetchSubmodules = true;
```

Comment:
@phiadaarr do you know why the submodule vendored dependency can't be removed with `-DNB_USE_SUBMODULE_DEPS=OFF` and adding `pkgs.robin-map` to `buildInputs` like so?

```nix
  pname = "nanobind";
  version = "1.9.0";
  pyproject = true;

  src = fetchFromGitHub {
    owner = "wjakob";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-NB14GMWKTMgxMLpMwF44fNW0fyjACSJxPhEp/xKYvnE=";
    fetchSubmodules = false;
  };

  disabled = pythonOlder "3.8";

  nativeBuildInputs = [ cmake ninja scikit-build setuptools ];
  buildInputs = [ eigen robin-map ];
  dontUseCmakeBuildDir = true;
  cmakeFlags = [
    "-DNB_USE_SUBMODULE_DEPS=OFF"
  ];
```

when I try compiling [pypcode](https://discourse.nixos.org/t/cmake-3-27-8-nix-support-setup-hook-line-126-cmake-command-not-found/39053/2) against the above it gives me header warnings such as

```text
In file included from /nix/store/h52r8rx5rdxak3ij9v5zz6rh8gvw2phv-python3.11-nanobind-1.9.0/lib/python3.11/site-packages/nanobind/src/nb_func.cpp:10:
/nix/store/h52r8rx5rdxak3ij9v5zz6rh8gvw2phv-python3.11-nanobind-1.9.0/lib/python3.11/site-packages/nanobind/src/nb_internals.h:14:10: fatal error: nanobind/nanobind.h: No such file or directory
   14 | #include <nanobind/nanobind.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Instead, I have to place `robin-map` where it expects the submodule.

```nix
  pname = "nanobind";
  version = "1.9.0";
  pyproject = true;

  src = fetchFromGitHub {
    owner = "wjakob";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-NB14GMWKTMgxMLpMwF44fNW0fyjACSJxPhEp/xKYvnE=";
    fetchSubmodules = false;
  };

  disabled = pythonOlder "3.8";

  nativeBuildInputs = [ cmake ninja scikit-build setuptools ];
  buildInputs = [ eigen robin-map ];
  dontUseCmakeBuildDir = true;
  preConfigure = ''
    mkdir -p ext/robin_map
    cp -r ${robin-map}/include -t ext/robin_map
  '';
```

Reply:
Sorry for the late response. As you have correctly figured out, the build system of nanobind expects `robin_map` to be located within `ext/`. If you put it somewhere else, the build will fails.

Reply:
@phiadaarr  I'm trying to avoid the vendored `robin_map` dependency pulled in with `fetchSubmodules = true`. In [`CMakeLists.txt`](https://github.com/wjakob/nanobind/blob/55bc50a797b5578d724701c97a8b647ca3b6f2bc/CMakeLists.txt#L15) there's a configuration option `NB_USE_SUBMODULE_DEPS` which controls whether nanobind should "Use the nanobind dependencies shipped as a git submodule of this repository". If this option is turned off, then [`cmake/nanobind-config.cmake`](https://github.com/wjakob/nanobind/blob/55bc50a797b5578d724701c97a8b647ca3b6f2bc/cmake/nanobind-config.cmake#L206-L219) searches for `robin_map` with `find_dependency(tsl-robin-map)`.

```cmake
# If nanobind was installed without submodule dependencies, then the
# dependencies directory won't exist and we need to find them.
# However, if the directory _does_ exist, then the user is free to choose
# whether nanobind uses them (based on `NB_USE_SUBMODULE_DEPS`), with a
# preference to choose them if `NB_USE_SUBMODULE_DEPS` is not defined
if (NOT IS_DIRECTORY ${NB_DIR}/ext/robin_map/include OR
    (DEFINED NB_USE_SUBMODULE_DEPS AND NOT NB_USE_SUBMODULE_DEPS))
  include(CMakeFindDependencyMacro)
  find_dependency(tsl-robin-map)
  target_link_libraries(${TARGET_NAME} PRIVATE tsl::robin_map)
else()
  target_include_directories(${TARGET_NAME} PRIVATE
    ${NB_DIR}/ext/robin_map/include)
endif()
```

That's why I'm confused why just adding `robin-map` to `buildInputs` and `"-DNB_USE_SUBMODULE_DEPS=OFF"` to `cmakeFlags` doesn't just work. The compilation succeeds but when I try to build [pypcode](https://discourse.nixos.org/t/cmake-3-27-8-nix-support-setup-hook-line-126-cmake-command-not-found/39053/2) e.g. against nanobind I get header not found errors. I'm not sure how one is supposed to feed dependencies to CMake otherwise.

Of course this discussion is mostly theoretical as one can just put `robin_map` where nanobind expects the submodule to avoid the CMake machinery which works and also avoids the vendored dependency.

```nix
  preConfigure = ''
    mkdir -p ext/robin_map
    cp -r ${robin-map}/include -t ext/robin_map
  '';
```

I asked the question in case there was something simple I was missing as I'm new to CMake. But concretely, with respect to the PR, is there a reason you didn't want to avoid the vendored dependency with the above snippet?

Reply:
Hi @phiadaarr ,

👋 Author here  ;-). Indeed, it would be good if you could leave this untouched since it affects ABI stability. If you have a Python process that imports multiple nanobind-based packages, the nanobind lib in each extension will talk to the others, and they will exchange internal data structures to enable inter-module data structure exchange. The robin_map data structure is part of this ABI contract. I am not saying that this is likely, but two identical nanobind versions referencing incompatible robin_map versions could hypothetically lead to undefined behavior.

Reply:
Thanks for making a nix package by the way!

Reply:
> Since the author of nanobind chose to pin an exact version of robin_map via a submodule, I believe we should follow this for nixpkgs and check out the exact version the author wants us to use.

This is subjective, but my impression of the philosophy of nixpkgs was to remove vendored dependencies wherever possible---as nixpkgs is a means of dependency specification itself, to keep software up to date, and to avoid the maintenance burden of keeping multiple versions around. For example, every python package ignores the canonical ways of specifying dependencies (`requirements.txt`, `pyproject.toml`, etc.) by necessity since multiple versions of the same python package conflict. From a user perspective, this decision makes tracking and manipulating dependencies (e.g. overriding, overlaying) transparent. Of course, the trade-off is that updates to dependencies can break parents, but generally nixpkgs skews heavily in favor of keeping software up to date.

On more practical note, the [pinned version](https://github.com/Tessil/robin-map/tree/68ff7325b3898fca267a103bad5c509e8861144d) is exactly the [latest version](https://github.com/Tessil/robin-map/releases/tag/v1.2.1) on nixpkgs and future breakage seems unlikely. However, I also don't see a compelling reason to overlay robin-map, so I agree that this is not a blocker.

> Sorry that I am not of much help here.

No worries! Apologies myself for the noise :p

Reply:
@stephen-huan did you see the example situation where a vendored dependency could potentially lead to ABI breakage and crashes in combination with other packages that haven't been built with nix?

Reply:
I wrote my comment before I saw yours, sorry. Your reasoning makes sense.

Reply:
Thanks for jumping in @wjakob! I think we can consider this resolved.